File: backend/internal/crypto/x25519.go
package crypto

import (
	"crypto/rand"
	"encoding/base64"
	"fmt"
	"strings"

	"golang.org/x/crypto/curve25519"
)

// GenerateKeyPair generates an X25519 key pair.
// Returns (publicKey, privateKey) as 32-byte slices.
func GenerateKeyPair() (publicKey, privateKey []byte, err error) {
	privateKey = make([]byte, curve25519.ScalarSize)
	if _, err = rand.Read(privateKey); err != nil {
		return nil, nil, fmt.Errorf("generate private key: %w", err)
	}

	// Clamp per RFC 7748
	privateKey[0] &= 248
	privateKey[31] &= 127
	privateKey[31] |= 64

	publicKey, err = curve25519.X25519(privateKey, curve25519.Basepoint)
	if err != nil {
		return nil, nil, fmt.Errorf("derive public key: %w", err)
	}

	return publicKey, privateKey, nil
}

// EncryptForRecipient encrypts a symmetric key for a recipient using their X25519 public key.
// Returns "base64(ephemeral_pub):base64(nonce):base64(encrypted_key)"
func EncryptForRecipient(recipientPublicKey, symmetricKey []byte) (string, error) {
	ephemeralPub, ephemeralPriv, err := GenerateKeyPair()
	if err != nil {
		return "", fmt.Errorf("generate ephemeral key: %w", err)
	}

	sharedSecret, err := curve25519.X25519(ephemeralPriv, recipientPublicKey)
	if err != nil {
		return "", fmt.Errorf("compute shared secret: %w", err)
	}

	encrypted, err := Encrypt(sharedSecret, symmetricKey)
	if err != nil {
		return "", fmt.Errorf("encrypt symmetric key: %w", err)
	}

	return base64.RawStdEncoding.EncodeToString(ephemeralPub) + ":" + encrypted, nil
}

// DecryptFromSender decrypts a symmetric key encrypted by EncryptForRecipient.
// encoded format: "base64(ephemeral_pub):base64(nonce):base64(encrypted_key)"
func DecryptFromSender(recipientPrivateKey []byte, encoded string) ([]byte, error) {
	idx := strings.Index(encoded, ":")
	if idx < 0 {
		return nil, fmt.Errorf("invalid encrypted key format")
	}

	ephemeralPubB64 := encoded[:idx]
	nonceCipher := encoded[idx+1:]

	ephemeralPub, err := base64.RawStdEncoding.DecodeString(ephemeralPubB64)
	if err != nil {
		return nil, fmt.Errorf("decode ephemeral public key: %w", err)
	}

	sharedSecret, err := curve25519.X25519(recipientPrivateKey, ephemeralPub)
	if err != nil {
		return nil, fmt.Errorf("compute shared secret: %w", err)
	}

	return Decrypt(sharedSecret, nonceCipher)
}

File: cli/src/crypto/x25519.rs
use anyhow::{anyhow, Context, Result};
use base64::{engine::general_purpose::STANDARD as B64, Engine};
use x25519_dalek::{PublicKey, StaticSecret};
use zeroize::Zeroizing;

use crate::crypto::chacha;

/// Encrypt a symmetric key for a recipient using their X25519 public key.
/// Returns "base64(ephemeral_pub):base64(nonce):base64(ciphertext)"
pub fn encrypt_for_recipient(recipient_pub_bytes: &[u8], sym_key: &[u8]) -> Result<String> {
    let recipient_pub_arr: [u8; 32] = recipient_pub_bytes
        .try_into()
        .map_err(|_| anyhow!("recipient public key must be 32 bytes"))?;
    let recipient_pub = PublicKey::from(recipient_pub_arr);

    // Generate ephemeral key pair
    let mut rng = rand::rngs::OsRng;
    let ephemeral_secret = StaticSecret::random_from_rng(&mut rng);
    let ephemeral_pub = PublicKey::from(&ephemeral_secret);

    // ECDH
    let shared = ephemeral_secret.diffie_hellman(&recipient_pub);
    let shared_bytes: [u8; 32] = *shared.as_bytes();
    let shared_key = Zeroizing::new(shared_bytes);

    // Encrypt sym key with shared secret
    let sym_key_arr: [u8; 32] = sym_key
        .try_into()
        .map_err(|_| anyhow!("sym key must be 32 bytes"))?;
    let encrypted = chacha::encrypt(&shared_key, &sym_key_arr).context("encrypt sym key")?;

    Ok(format!("{}:{}", B64.encode(ephemeral_pub.as_bytes()), encrypted))
}

/// Decrypt a symmetric key encrypted by encrypt_for_recipient.
/// encoded format: "base64(ephemeral_pub):base64(nonce):base64(ciphertext)"
pub fn decrypt_sym_key(
    recipient_priv_bytes: &[u8; 32],
    encoded: &str,
) -> Result<Zeroizing<Vec<u8>>> {
    let first_colon = encoded
        .find(':')
        .ok_or_else(|| anyhow!("invalid encrypted sym key format"))?;

    let ephemeral_pub_b64 = &encoded[..first_colon];
    let nonce_cipher = &encoded[first_colon + 1..];

    let ephemeral_pub_bytes: [u8; 32] = B64
        .decode(ephemeral_pub_b64)
        .context("decode ephemeral pub")?
        .try_into()
        .map_err(|_| anyhow!("ephemeral pub must be 32 bytes"))?;

    let our_secret = StaticSecret::from(*recipient_priv_bytes);
    let their_pub = PublicKey::from(ephemeral_pub_bytes);
    let shared = our_secret.diffie_hellman(&their_pub);
    let shared_arr: [u8; 32] = *shared.as_bytes();
    let shared_key = Zeroizing::new(shared_arr);

    let plaintext = chacha::decrypt(&shared_key, nonce_cipher)
        .context("decrypt sym key")?;

    Ok(Zeroizing::new(plaintext))
}
